### PR TITLE
modify the strtof

### PR DESCRIPTION
**Pull request by flyingfish89**
## Summary
Add function:decfloat(fix the bug of the wrong return number 2147483648 and some wrong return like 2147483520.0)
                      hexfloat(it's copy from musl libc)
## Impact
 modify the strtof and strtod
## Testing
testing from wasm

### CODE REVIEW

File: libs/libc/stdlib/lib_strtod.c [hunk 1]
```diff
@@ -67,28 +67,105 @@
 #  define __DBL_MAX_EXP__ (1024)
 #endif
 
+#define shgetc(f) (*f++)
+#define shunget(f) (f--)
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtod.c:71]:
```suggestion
#define shgetc(f) (*(f)++)
#define shunget(f) ((f)--)
```

File: libs/libc/stdlib/lib_strtod.c [hunk 2]
```diff
@@ -67,28 +67,105 @@
...
+#define ifexist(A,B) if(*A) {**A = B;}
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtod.c:72]:
```suggestion
#define ifexist(A,B) if(*(A) != NULL) {**(A) = (B);}
```

File: libs/libc/stdlib/lib_strtod.c [hunk 3]
```diff
@@ -67,28 +67,105 @@
...
+ *
+ * Description:
+ *   Gets the sum number of a string which the value of
+ *   each character between 0 and 9
+ *
+ *   flag:1 , the value of the f will be change ;0 ,not change
+ *
+ ****************************************************************************/
+
+static long long scanexp (char** f , bool flag)
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtod.c:89]: please add `FAR` and remove spaces before `(` in function prototype

File: libs/libc/stdlib/lib_strtod.c [hunk 4]
```diff
@@ -67,28 +67,105 @@
...
+    {
+      ifexist(&f, s);
+    }
+
+  if (isdigit(c))
+    {
+      return false;
+    }
+
+  return true;
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtod.c:154]:
```suggestion

  return !isdigit(c);
```

File: libs/libc/stdlib/lib_strtod.c [hunk 5]
```diff
@@ -109,235 +186,423 @@ static bool chtod(char c, double base, FAR double *number)
...
+  long long num_decimal;
+  long double y;
 
-  /* Handle optional sign */
+  const double zero = 0.;
+  const int p10s[] =
+              {
+                10, 100, 1000, 10000, 100000,
+                1000000, 10000000, 100000000
+              };
```
Comment (pkarashchenko) [inline libs/libc/stdlib/lib_strtod.c:219]: remove some spaces

File: libs/libc/stdlib/lib_strtod.c [hunk 6]
```diff
@@ -67,28 +67,100 @@
...
+/****************************************************************************
+ * Name: chtou
+ *
+ * Description:
+ *   Determine whether c is a keyword Set the
+ *   value of the number add base
+ *
+ ****************************************************************************/
+
+static bool chtou(char c, int base, uint32_t *number)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:161]:
```suggestion
static bool chtou(char c, int base, FAR uint32_t *number)
```

File: libs/libc/stdlib/lib_strtod.c [hunk 7]
```diff
@@ -109,235 +181,462 @@ static bool chtod(char c, double base, FAR double *number)
...
- *   - Hexadecimal input is not supported
- *   - INF, INFINITY, NAN, and NAN(...) are not supported
+ *   Implement functionality similar to scalbnl
+ *   return a long number of (number * base ^ exp)
  *
  ****************************************************************************/
 
-double strtod(FAR const char *str, FAR char **endptr)
+static long double scalbnx(long double number,
+  long double base, long long exp)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:198]:
```suggestion
                            long double base, long long exp)
```

File: libs/libc/stdlib/lib_strtod.c [hunk 8]
```diff
@@ -109,235 +181,462 @@ static bool chtod(char c, double base, FAR double *number)
...
- *   NOTE: This implementation is limited as compared to POSIX:
- *   - Hexadecimal input is not supported
- *   - INF, INFINITY, NAN, and NAN(...) are not supported
+ *   Implement functionality similar to scalbnl
+ *   return a long number of (number * base ^ exp)
  *
  ****************************************************************************/
 
-double strtod(FAR const char *str, FAR char **endptr)
+static long double scalbnx(long double number,
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:197]: only exist when CONFIG_HAVE_LONG_DOUBLE is defined

File: libs/libc/stdlib/lib_strtod.c [hunk 9]
```diff
@@ -109,235 +181,462 @@ static bool chtod(char c, double base, FAR double *number)
...
- * Name: strtod
+ * Name: scalbnx
  *
  * Description:
- *   Convert a string to a double value
- *
- *   NOTE: This implementation is limited as compared to POSIX:
- *   - Hexadecimal input is not supported
- *   - INF, INFINITY, NAN, and NAN(...) are not supported
+ *   Implement functionality similar to scalbnl
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:192]: if so, why not call scalbnl directly

File: libs/libc/stdlib/lib_strtod.c [hunk 10]
```diff
@@ -109,235 +181,462 @@ static bool chtod(char c, double base, FAR double *number)
...
+
+/****************************************************************************
+ * Name: decfloat
+ *
+ * Description:
+ *   Convert a decimal string to a float value
+ *
+ ****************************************************************************/
+
+double decdouble(FAR char *ptr, FAR char ***endptr)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:235]: add static

File: libs/libc/stdlib/lib_strtod.c [hunk 11]
```diff
@@ -109,235 +181,462 @@ static bool chtod(char c, double base, FAR double *number)
...
+/****************************************************************************
+ * Name: hexdouble
+ *
+ * Description:
+ *   Convert a hexadecimal string to a double value
+ *
+ ****************************************************************************/
 
-  if (negative)
+double hexdouble(FAR char *ptr, FAR char ***endptr)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:382]: add static

File: libs/libc/stdlib/lib_strtod.c [hunk 12]
```diff
@@ -109,235 +181,462 @@ static bool chtod(char c, double base, FAR double *number)
...
-      p10 *= p10;
+  default:
+      break;
     }
 
-  if (!is_real(number))
+  for (i = 0; i < 8 && (*s | 32) == "infinity"[i]; i++)
     {
-      set_errno(ERANGE);
+      if (s) s++;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:607]: if (s)
  {
    s++
  }

File: libs/libc/stdlib/lib_strtod.c [hunk 13]
```diff
@@ -109,235 +181,462 @@ static bool chtod(char c, double base, FAR double *number)
...
+      ifexist((&endptr), s);
+      return negative ? -INFINITY : INFINITY;
     }
 
-  return number;
-}
+  s -= i;
+  for (i = 0; i < 3 && (*s | 32) == "nan"[i]; i++)
+    {
+      if (s) s++;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:619]: ditto

File: libs/libc/stdlib/lib_strtod.c (hunk 9 above)
Comment (flyingfish89) [reply libs/libc/stdlib/lib_strtod.c:192]: because the avr-libc has no functions scalbnl and powl

Comment (xiaoxiang781216) [reply libs/libc/stdlib/lib_strtod.c:198]: but should we use https://github.com/apache/nuttx/blob/master/libs/libc/math/lib_scalbnl.c?

File: libs/libc/stdlib/lib_strtod.c [hunk 14]
```diff
@@ -67,28 +67,106 @@
...
 
+#ifdef CONFIG_HAVE_LONG_DOUBLE
+#define long_double long double
+#else
+#define long_double double
+#endif
+
+#define shgetc(f) (*(f)++)
+#define shunget(f) ((f)--)
+#define ifexist(a,b) if((*a) != NULL) {**(a) = (b);}
```
Comment (hartmannathan) [inline libs/libc/stdlib/lib_strtod.c:78]:
```suggestion
#define ifexist(a,b) do { if ((*a) != NULL) {**(a) = (b);} } while (0)
```

File: libs/libc/stdlib/lib_strtof.c [hunk 15]
```diff
@@ -69,28 +71,108 @@
...
+#define long_double long double
+#elif CONFIG_HAVE_DOUBLE
+#define long_double double
+#else
+#define long_double float
+#endif
+
+#define shgetc(f) (*(f)++)
+#define shunget(f) ((f)--)
+#define ifexist(a,b) if((*a) !=NULL)  {**(a) = (b);}
```
Comment (hartmannathan) [inline libs/libc/stdlib/lib_strtof.c:84]:
```suggestion
#define ifexist(a,b) do { if ((*a) != NULL) {**(a) = (b);} } while (0)
```

Perhaps the duplicated macro(s) can be moved to a common header?

File: libs/libc/stdlib/lib_strtof.c [hunk 16]
```diff
@@ -69,28 +74,108 @@
 #  define __FLT_MAX_EXP__ (128)
 #endif
 
+#ifdef CONFIG_HAVE_LONG_DOUBLE
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtof.c:77]: if you convert all computation to long double, why not forward strtof to strod directly like musl?

**Comment by xiaoxiang781216** [comment]
the patch need rebase again

File: libs/libc/stdlib/lib_strtod.c [hunk 17]
```diff
@@ -109,235 +190,518 @@ static bool chtod(char c, double base, FAR double *number)
...
+
+#ifdef CONFIG_HAVE_DOUBLE
+
+double inline strtod(FAR const char *str, FAR char **endptr)
+{
+  return strtox(str, endptr, 2);
 }
 
-#endif /* CONFIG_HAVE_DOUBLE */
+#endif /* CONFIG_HAVE_DOUBLE */
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:707]: add newline

File: libs/libc/stdlib/lib_strtod.c [hunk 18]
```diff
@@ -109,235 +190,518 @@ static bool chtod(char c, double base, FAR double *number)
...
+ * Name: strtod
+ *
+ * Description:
+ *   Convert a string to a double value
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_HAVE_DOUBLE
+
+double inline strtod(FAR const char *str, FAR char **endptr)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtod.c:702]: what' about strtold

File: libs/libc/stdlib/lib_strtold.c [hunk 19]
```diff
@@ -55,234 +62,666 @@
...
+ * Name: strtold
+ *
+ * Description:
+ *   Convert a string to a long double value
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_HAVE_LONG_DOUBLE
+
+long double inline strtold(FAR const char *str, FAR char **endptr)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:722]: remove inline

File: libs/libc/stdlib/lib_strtold.c [hunk 20]
```diff
@@ -55,234 +62,666 @@
...
+ * Name: strtod
+ *
+ * Description:
+ *   Convert a string to a double value
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_HAVE_DOUBLE
+
+double inline strtod(FAR const char *str, FAR char **endptr)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:705]: remove inline

File: libs/libc/stdlib/lib_strtold.c [hunk 21]
```diff
@@ -55,234 +62,666 @@
...
+
+/****************************************************************************
+ * Name: strtof
+ *
+ * Description:
+ *   Convert a string to a float value
+ *
+ ****************************************************************************/
+
+float inline strtof(FAR const char *str, FAR char **endptr)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:690]: remove inline

File: libs/libc/stdlib/lib_strtold.c [hunk 22]
```diff
@@ -55,234 +62,666 @@
...
+ * Name: strtox
+ *
+ * Description:
+ *   Convert a string to a long_double value
+ *
+ ****************************************************************************/
+
+static long_double strtox(FAR const char *str, FAR char ***endptr, int flag)
+{
+  FAR char *s = (FAR char *) str;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:588]: remove space before str

File: libs/libc/stdlib/lib_strtold.c [hunk 23]
```diff
@@ -55,234 +62,666 @@
...
-            }
+          s++;
         }
+    }
 
-      n >>= 1;
-      p10 *= p10;
+  if (i == 3 || i == 8)
+    {
+      ifexist((&endptr), s);
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:644]:
```suggestion
      ifexist(&endptr, s);
```

File: libs/libc/stdlib/lib_strtold.c [hunk 24]
```diff
@@ -55,234 +62,666 @@
...
+  case '+':
+      s++;
+
+  default:
+      break;
+    }
+
+  for (i = 0; i < 8 && (*s | 32) == "infinity"[i]; i++)
+    {
+      if (s)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:636]: why need check s

File: libs/libc/stdlib/lib_strtold.c [hunk 25]
```diff
@@ -55,234 +62,666 @@
...
+      ifexist((&endptr), s);
+      return negative ? -INFINITY : INFINITY;
     }
 
-  if (!is_real(number))
+  s -= i;
+  for (i = 0; i < 3 && (*s | 32) == "nan"[i]; i++)
     {
-      set_errno(ERANGE);
+      if (s)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:651]: why need check s

File: libs/libc/stdlib/lib_strtold.c [hunk 26]
```diff
@@ -55,234 +62,666 @@
...
+          s++;
+        }
     }
 
-errout:
-  if (endptr)
+  if (i == 3)
     {
-      *endptr = p;
+      ifexist((&endptr), s);
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:659]:
```suggestion
      ifexist(&endptr, s);
```

File: libs/libc/stdlib/lib_strtold.c [hunk 27]
```diff
@@ -55,234 +62,666 @@
...
+
+/****************************************************************************
+ * Name: strtox
+ *
+ * Description:
+ *   Convert a string to a long_double value
+ *
+ ****************************************************************************/
+
+static long_double strtox(FAR const char *str, FAR char ***endptr, int flag)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:586]: why endptr type si char *** not char **

File: libs/libc/stdlib/lib_strtold.c [hunk 28]
```diff
@@ -55,234 +62,666 @@
...
+    {
+      s += 2;
+      y = hexfloat(s, &endptr, bits, emin);
+    }
+  else
+    {
+      y = decfloat(s, &endptr);
+    }
+
+  return negative? -y : y;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:675]:
```suggestion
  return negative ? -y : y;
```

File: libs/libc/stdlib/lib_strtold.c [hunk 29]
```diff
@@ -55,234 +62,666 @@
...
+ * Name: hexfloat
+ *
+ * Description:
+ *   Convert a hexadecimal string to a float value
+ *
+ ****************************************************************************/
 
-  if (negative)
+static long_double hexfloat(FAR char *ptr,
+                            FAR char ****endptr, int bits, int emin)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:393]:
```suggestion
                            FAR char **endptr, int bits, int emin)
```

File: libs/libc/stdlib/lib_strtold.c [hunk 30]
```diff
@@ -55,234 +62,666 @@
...
+{
+  FAR char *f = (FAR char *)ptr;
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:395]: remove cast

File: libs/libc/stdlib/lib_strtold.c [hunk 31]
```diff
@@ -55,234 +62,666 @@
...
+          x = 0;
         }
+    }
 
-      exponent -= num_decimals;
+  if (num_digit <= 9 && num_decimal == 0)
+    {
+      return (long_double)x;
+    }
+  else if (num_digit + num_decimal > LDBL_MAX_10_EXP)
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:365]: you can't use LDBL_MAX_10_EXP directly

File: libs/libc/stdlib/lib_strtold.c [hunk 32]
```diff
@@ -55,234 +52,749 @@
...
+
+  if (i == 3)
+    {
+      ifexist(endptr, s);
+      return NAN;
+    }
+
+  /* Process optional 0x prefix */
+
+  if (*s == '0' && (*(s + 1) | 32) == 'x' )
```
Comment (xiaoxiang781216) [inline libs/libc/stdlib/lib_strtold.c:717]:
```suggestion
  if (*s == '0' && (*(s + 1) | 32) == 'x')
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by fjpanag** [comment]
This PR broke our tests.

Specifically the Lua test fails on the following line:

```lua
assert(string.len(string.format('%99.99f', -1e308)) >= 100)
```

`string.format('%99.99f', -1e308)` returns the following:

```
                                                                                               -inf
```

Before this PR, the output was:

```
-100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
```

**Comment by flyingfish89** [comment]
> This PR broke our tests.
> 
> Specifically the Lua test fails on the following line:
> 
> ```lua
> assert(string.len(string.format('%99.99f', -1e308)) >= 100)
> ```
> 
> `string.format('%99.99f', -1e308)` returns the following:
> 
> ```
>                                                                                                -inf
> ```
> 
> Before this PR, the output was:
> 
> ```
> -100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
> ```
> 
> **EDIT:** I tried the same on my computer. I am running Lua 5.4.4 on Fedora 37. The output now is:
> 
> ```
> -100000000000000001097906362944045541740492309677311846336810682903157585404911491537163328978494688899061249669721172515611590283743140088328307009198146046031271664502933027185697489699588559043338384466165001178426897626212945177628091195786707458122783970171784415105291802893207873272974885715430223118336.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
> ```

I  have tried this test and return is 
捕获

**Comment by flyingfish89** [comment]
捕获

**Comment by flyingfish89** [comment]
Have you configured CONFIG_HAVE_DOUBLE

**Comment by hartmannathan** [comment]
If not CONFIG_HAVE_DOUBLE then should there be some adjustment to the strtox function to deal with different overflow levels etc?

**Comment by flyingfish89** [comment]
> If not CONFIG_HAVE_DOUBLE then should there be some adjustment to the strtox function to deal with different overflow levels etc?

em....,if not config_have_double ,the return is -inf when input is -1e138

**Comment by flyingfish89** [comment]
> This PR broke our tests.
> 
> Specifically the Lua test fails on the following line:
> 
> ```lua
> assert(string.len(string.format('%99.99f', -1e308)) >= 100)
> ```
> 
> `string.format('%99.99f', -1e308)` returns the following:
> 
> ```
>                                                                                                -inf
> ```
> 
> Before this PR, the output was:
> 
> ```
> -100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
> ```
> 
> **EDIT:** I tried the same on my computer. I am running Lua 5.4.4 on Fedora 37. The output now is:
> 
> ```
> -100000000000000001097906362944045541740492309677311846336810682903157585404911491537163328978494688899061249669721172515611590283743140088328307009198146046031271664502933027185697489699588559043338384466165001178426897626212945177628091195786707458122783970171784415105291802893207873272974885715430223118336.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
> ```

could you describe how this scenario would have been triggered

**Comment by fjpanag** [comment]
> CONFIG_HAVE_DOUBLE

Yes `CONFIG_HAVE_DOUBLE` is defined (although as I see, there is no Kconfig for it).

> could you describe how this scenario would have been triggered

If you mean any real-life scenario, I have no idea.  
This is part of the standard Lua test suite, and they include this specific check.

It is there either because it is needed by the VM itself in a way that I do not know, or to check conformance of the libraries to the standard.

**Comment by flyingfish89** [comment]
> > CONFIG_HAVE_DOUBLE
> 
> Yes `CONFIG_HAVE_DOUBLE` is defined (although as I see, there is no Kconfig for it).
> 
> > could you describe how this scenario would have been triggered
> 
> If you mean any real-life scenario, I have no idea. This is part of the standard Lua test suite, and they include this specific check.
> 
> It is there either because it is needed by the VM itself in a way that I do not know, or to check conformance of the libraries to the standard.

em.....,so what board should I test it on and what configuration should I turn it on.

**Comment by xiaoxiang781216** [comment]
@fjpanag could you share your config or add the test to sim:lua?

**Comment by fjpanag** [comment]
>  em.....,so what board should I test it on and what configuration should I turn it on.

Ah, sorry I misunderstood your question!

I am running these tests in sim.
I've attached my defconfig (app-specific configs are removed).

[sim-defconfig.txt](https://github.com/apache/nuttx/files/10578244/sim-defconfig.txt)


In the meantime I will try to replicate this on `sim:lua` config.

**Comment by fjpanag** [comment]
I just tried using the official NuttX app:

```sh
$ git clone https://github.com/apache/nuttx.git nuttx
$ git clone https://github.com/apache/nuttx-apps.git apps
$ cd nuttx
$ ./tools/configure.sh sim:lua
$ make all
```

And then:

```
$ ./nuttx

NuttShell (NSH) NuttX-12.0.0
nsh> lua
Lua 5.4.0  Copyright (C) 1994-2020 Lua.org, PUC-Rio
> string.format('%99.99f', -1e308)
                                                                                               -inf
> 
```

**Comment by hartmannathan** [comment]
Perhaps it would be beneficial to have an automated test suite for our libc, in which functions like strtod() (and the reverse) would be tested. We'd have a list of inputs and their known-correct outputs and compare those. Also that would serve as documentation of what the expected results are. If there are changes in the future that alter the results, we'll know right away.

**Comment by flyingfish89** [comment]
Can you change strtold.c to look like this, and then test if there are any other problems,thanks!
捕获
@fjpanag

**Comment by fjpanag** [comment]
@flyingfish89 I just applied this, and the tests now pass.

**Comment by flyingfish89** [comment]
> @flyingfish89 I just applied this, and the tests now pass.

ok.thanks!

**Comment by acassis** [comment]
> Perhaps it would be beneficial to have an automated test suite for our libc, in which functions like strtod() (and the reverse) would be tested. We'd have a list of inputs and their known-correct outputs and compare those. Also that would serve as documentation of what the expected results are. If there are changes in the future that alter the results, we'll know right away.

@linguini1 what you think about this idea? To include libc testing like this in NUTS ?

**Comment by jerpelea** [comment]
PR merged without proper PR title 
ex:
libs/libc: modify the strtof

**Comment by linguini1** [comment]
> @linguini1 what you think about this idea? To include libc testing like this in NUTS ?

I was under the impression that NuttX had a libc test suite already? I think it should be separate from NUTS, since testing libc is a common task and I think there are a lot of existing test solutions we can re-use for that instead of re-inventing the wheel.

**Pull request merged**